Creates -datadir immediately on startup

nkinkade opened the pull request:
Currently, the value of the flag -datadir is only created when uuid-annotator goes to write the first annotation file. If uuid-annotator is running in a pod that doesn't get sufficient traffic which would cause tcp-info to generate an event, then the datatype directory will not be created. However, pusher will crashloop until this directory exists. This commit causes uuid-annotator to create the datatype directory immediately upon startup.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/uuid-annotator/56)
<!-- Reviewable:end -->

nkinkade committed:
Creates -datadir immediately on startup

Currently, the value of the flag -datadir is only created when
uuid-annotator goes to write the first annotation file. If
uuid-annotator is running in a pod that doesn't get sufficient traffic
which would cause tcp-info to generate an event, then the datatype
directory will not be created. However, pusher will crashloop until this
directory exists. This commit causes uuid-annotator to create the
datatype directory immediately upon startup.

stephen-soltesz reviewed (commented):
> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/m-lab/uuid-annotator/56)* status: 0 of 1 approvals obtained (waiting on @nkinkade)

___
*[`main.go` line 78 at r1](https://reviewable.io/reviews/m-lab/uuid-annotator/56#-N_dzBEx-McuX4BrrQl2:-N_dzBEx-McuX4BrrQl3:b19kqmy) ([raw file](https://github.com/m-lab/uuid-annotator/blob/3e47dcad3d35a3b2f0953617eb35bc05d61cbe57/main.go#L78)):*
> ```Go
> 	// Create the datatype directory immediately, since pusher will crash
> 	// without it.
> 	rtx.Must(os.MkdirAll(*datadir, 0755), "Could not create datatype dir %s", datadir)
> ```

With our goal of running containers with non-root privileges, is 0755 as the target directory mode correct? (It seems reasonable).

Please remove the second `MkdirAll` with a different mode. I don't think we need both now.  https://github.com/m-lab/uuid-annotator/blob/main/handler/handler.go#L37


<!-- Sent from Reviewable.io -->

nkinkade reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/m-lab/uuid-annotator/56)* status: 0 of 1 approvals obtained (waiting on @stephen-soltesz)

___
*[`main.go` line 78 at r1](https://reviewable.io/reviews/m-lab/uuid-annotator/56#-N_dzBEx-McuX4BrrQl2:-N_e-e4o1wfUg8MIzlOj:b39xqxt) ([raw file](https://github.com/m-lab/uuid-annotator/blob/3e47dcad3d35a3b2f0953617eb35bc05d61cbe57/main.go#L78)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

With our goal of running containers with non-root privileges, is 0755 as the target directory mode correct? (It seems reasonable).

Please remove the second `MkdirAll` with a different mode. I don't think we need both now.  https://github.com/m-lab/uuid-annotator/blob/main/handler/handler.go#L37
</blockquote></details>

I think 755 should be sufficient. uuid-annotator does not run as root. tcp-info uses 755 for the directory as well, and I think so does ndt-server. I noticed the 777 in handler.go and thought it wasn't 100% right but certainly not hurting anything. I can chance the call in handler.go to use 755, too, if you think it would be better.

I don't think we can remove the `MkdirAll()` in handler.go. That call to `MkdirAll()` doesn't just create the base datatype directory, but all the *year/month/day* subdirectories as well. That one should stay, I believe, and it just also happened to take care of creating the base datatype directory. Now it will only create the data sub-paths.


<!-- Sent from Reviewable.io -->

stephen-soltesz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/m-lab/uuid-annotator/56)* status: 0 of 1 approvals obtained (waiting on @nkinkade)

___
*[`main.go` line 78 at r1](https://reviewable.io/reviews/m-lab/uuid-annotator/56#-N_dzBEx-McuX4BrrQl2:-N_e1nie61gRQpWCdbjy:b2thg4g) ([raw file](https://github.com/m-lab/uuid-annotator/blob/3e47dcad3d35a3b2f0953617eb35bc05d61cbe57/main.go#L78)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

I think 755 should be sufficient. uuid-annotator does not run as root. tcp-info uses 755 for the directory as well, and I think so does ndt-server. I noticed the 777 in handler.go and thought it wasn't 100% right but certainly not hurting anything. I can chance the call in handler.go to use 755, too, if you think it would be better.

I don't think we can remove the `MkdirAll()` in handler.go. That call to `MkdirAll()` doesn't just create the base datatype directory, but all the _<year>/<month>/<day>_ subdirectories as well. That one should stay, I believe, and it just also happened to take care of creating the base datatype directory. Now it will only create the data sub-paths.
</blockquote></details>

I think consistency is preferable where possible. If we need a different mode in handler.go I'd like to understand why. If we don't, I'd like to use the same mode throughout.


<!-- Sent from Reviewable.io -->

stephen-soltesz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/m-lab/uuid-annotator/56)* status: 0 of 1 approvals obtained (waiting on @nkinkade)

___
*[`main.go` line 78 at r1](https://reviewable.io/reviews/m-lab/uuid-annotator/56#-N_dzBEx-McuX4BrrQl2:-N_e2BM55rFt3OqXEk2V:bwf65bz) ([raw file](https://github.com/m-lab/uuid-annotator/blob/3e47dcad3d35a3b2f0953617eb35bc05d61cbe57/main.go#L78)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

I think consistency is preferable where possible. If we need a different mode in handler.go I'd like to understand why. If we don't, I'd like to use the same mode throughout.
</blockquote></details>

(And, thank you for clarifying that handler.go needs to create all the subdirs -- I wasn't thinking about that.)


<!-- Sent from Reviewable.io -->

nkinkade committed:
Uses mode 755 for MkdirAll() in handler.go

777 is unnecessary, and using 755 is more consistent with the permission
on the base datatype directory, as well as more consistent with how
other sidecar services permission their directories.

nkinkade reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/m-lab/uuid-annotator/56)* status: 0 of 1 approvals obtained (waiting on @stephen-soltesz)

___
*[`main.go` line 78 at r1](https://reviewable.io/reviews/m-lab/uuid-annotator/56#-N_dzBEx-McuX4BrrQl2:-N_e5UxU5rv3wFJA6SkV:b-t4r2pb) ([raw file](https://github.com/m-lab/uuid-annotator/blob/3e47dcad3d35a3b2f0953617eb35bc05d61cbe57/main.go#L78)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

(And, thank you for clarifying that handler.go needs to create all the subdirs -- I wasn't thinking about that.)
</blockquote></details>

I've changed the call to `MkdirAll()` in handler.go use mode 0755 to be consistent with the parent directories.


<!-- Sent from Reviewable.io -->

stephen-soltesz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> -- Thank you!

> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/m-lab/uuid-annotator/56#-:-N_e5qOtE9xoZw8JN5Zs:bspa67d)* status: :shipit: complete! 1 of 1 approvals obtained



<!-- Sent from Reviewable.io -->
